### ISSUE TITLE
bug: selection of renamed columns is optimized incorrectly

### ISSUE THREAD

**Issue by John-Boik** [ibis-project/ibis#1045]
Hello.
In doing a double join to connect three tables, I've identified behavior of Ibis that I don't understand. I'm hoping that someone can shed light on why it's occurring, or how the error message relates to what is occurring. If I write the script one way, the double join works. If I change a few lines, the second join throws a "More than one base table not implemented" error. I'm using the postgres connection.

A test script is below. In it, I create some small test tables that mimic the data I'm working with. For example, some column types need to be recast, and some column names appear in more than one table. I rename some of these columns to avoid conflicts.  

In the script, if the "if True" is changed to False, the alternative script is run, which throws the error. Otherwise, the script works fine.  Any suggestions to improve this code snippet are also welcome. 

```
con = ibis.postgres.connect(host="localhost", database="...", user='...', password="...")
con.raw_sql("drop table if exists test1")
con.raw_sql("drop table if exists test2")
con.raw_sql("drop table if exists test3")

con.raw_sql("create table test1 (id1 integer not null, val1 real)")
for i in range(10):
  con.raw_sql("insert into test1 values (1, 10), (2, 20), (3, 20), (2, 21)")

con.raw_sql("create table test2 (id2a integer not null, id2b bigint, val2 double precision)")
for i in range(10):
  con.raw_sql("insert into test2 values (1, 4, 15), (2, 5, 25)")

con.raw_sql("create table test3 (id3 text not null, val2 double precision, dt timestamp without time zone)")
for i in range(10):
  con.raw_sql("insert into test3 values ('4', 15, '2016-07-17 19:55:03'), (2, 25, '2016-07-17 19:55:03')")

t1 = con.table('test1')
t2 = con.table('test2')
t3 = con.table('test3')

c = t3.id3.cast('int64').name('id3')
ls = t3.columns
ls.remove('id3')
t3 = t3[ls].mutate(c)

if True:
  """this alternative works fine""" 
  t3 = t3[[t3[i].name('t3_'+ i) for i in t3.columns]]
  print("\n first t3 info:")
  print(t3.info())
  
  cond1 = t3.t3_id3 == t2.id2b

else:
  """this alternative causes an error on the second join"""
  c = t3.id3.name('t3_val2')
  ls = t3.columns
  ls.remove('val2')
  t3 = t3[ls].mutate(c) 
  print("\n second t3 info:")
  print(t3.info()) 
  
  cond1 = t3.id3 == t2.id2b

t4 = t3.join(t2, cond1).materialize()  # first join

t1 = t1[[t1[i].name('t1_'+ i) for i in t1.columns]]

cond2 = t1.t1_id1 == t4.id2a
t5 = t1.join(t4, cond2).materialize()  # second join

print(t5.info())

```
The t3 table seems to be mostly the same with either alternative, other than some renamed columns.
The error raised on print(t5.info()) is:
  File ".../lib/python3.5/site-packages/ibis/expr/analysis.py", line 1056, in find_source_table
  raise NotImplementedError('More than one base table not implemented')
  NotImplementedError: More than one base table not implemented

**Comment by John-Boik** [comment]
Ah.  It looks like the second alternative works with the following change:
`t3 = t3[ls].mutate(c)`
to:
`t3 = t3[ls].mutate(c).view()`

**Comment by John-Boik** [comment]
As a suggestion, it might be helpful to those who run into the "NotImplementedError: More than one base table not implemented" error if the message provided more context, description, or hints as to what the problem is. In my example, the problem was not that a feature was not implemented, it was that a table was not considered a distinct relation for evaluation purposes.

**Comment by cpcloud** [comment]
@John-Boik Thanks for the detailed report! I agree with you that we should improve this error message. Slating this for 0.11.1.

**Comment by cpcloud** [comment]
@John-Boik This is now giving an error from postgres itself. Since there's a workaround available and this is a particularly thorny thing to fix I'm going to push it to the next release.

The main issue here is related to optimization.

The `t4` query is problematic:

```
ref_0
PostgreSQLTable[table]
  name: test3
  schema:
    val2 : string
    dt : double

ref_1
Selection[table]
  table:
    Table: ref_0
  selections:
    val2 = Column[double*] 'val2' from table
      ref_0
    dt = Column[timestamp*] 'dt' from table
      ref_0
    id3 = Cast[int64*]
      id3 = Column[string*] 'id3' from table
        ref_0
      to:
        int64

ref_2
Selection[table]
  table:
    Table: ref_1
  selections:
    dt = Column[timestamp*] 'dt' from table
      ref_1
    id3 = Column[int64*] 'id3' from table
      ref_1
    t3_val2 = Column[int64*] 'id3' from table
      ref_1

ref_3
PostgreSQLTable[table]
  name: test2
  schema:
    id2a : int32
    id2b : int64
    val2 : double

InnerJoin[table]
  left:
    Table: ref_2
  right:
    Table: ref_3
  predicates:
    Equals[boolean*]
      left:
        id3 = Column[int64*] 'id3' from table
          ref_1
      right:
        id2b = Column[int64*] 'id2b' from table
          ref_3
```

The problematic part is the join:

```
InnerJoin[table]
  left:
    Table: ref_2
  right:
    Table: ref_3
  predicates:
    Equals[boolean*]
      left:
        id3 = Column[int64*] 'id3' from table
          ref_1
      right:
        id2b = Column[int64*] 'id2b' from table
          ref_3
```

`id3` is being taken from `ref_1`, when it should really be coming from `ref_2`. Somewhere in the codebase we're trying to pull up projections into the root expression in a way that is algebraically incorrect.

Column expressions cannot be pulled up if there are any renames happening at all, since a rename implies a distinct relation whose columns must be referred to off of that relation.

I'll open a new issue about this problem and leave this issue open until we have a chance to fix it.

**Comment by cpcloud** [comment]
There's a much smaller reprex here:

```python
        t2 = ibis.table(
            [("id2a", "int32"), ("id2b", "int64"), ("val2", "float64")],
            name="table2",
        )
        t3 = ibis.table([("id3", "int32"), ("val2", "float64")], name="table3")
        t3 = t3.drop("val2")
        expr = t3.join(t2, t3.id3 == t2.id2b)
```

It's the drop + join that is the underlying problem here, and a `drop` is just a special case of a projection.

Ibis is again trying to optimize and fails to account for the fact that you cannot hand sqlalchemy relations at different scopes without it generating undesirable cross joins everywhere.

**Comment by cpcloud** [comment]
The reprex was fixed by ba4e5e56f7e0ccd367a270d6d5090f340d610dae, but the join with a join on the right side (`t1.join(t4)`) is still broken. We'll tackle this in 3.x. I'll open an issue now to track that.
